### PR TITLE
SR-12022: refactor LiteralExpr to combine common initializer code

### PR DESCRIPTION
**Pull request by Interfere**
Revisit inheritance chain for `LiteralExpr`.
* Hoist `get/setInitializer()` methods to abstract class `LiteralExpr`;
* Introduce new abstract class `BuiltinLiteralExpr` and hoist `get/setBuiltinInitializer` methods to it;
* Simplify implementation of `SILGenFunction::emitLiteral()`;
* Rename `get/setResultInit()` ~> `get/setInitializer()`;
* Update `include/swift/AST/ExprNodes.def`.

Resolves SR-12022.

### CODE REVIEW

**Comment by Interfere** [comment]
@brentdax could you please review?

**Review by xedin — APPROVED** [review]
Type-checker changes look good to me!

File: lib/SILGen/SILGenApply.cpp [hunk 1]
```diff
@@ -1786,6 +1786,139 @@ static bool hasUnownedInnerPointerResult(CanSILFunctionType fnType) {
...
+    return buildBuiltinLiteralArgs(nilLiteral);
+  } else if (auto booleanLiteral = dyn_cast<BooleanLiteralExpr>(literal)) {
+    return buildBuiltinLiteralArgs(SGF, C, booleanLiteral);
+  } else if (auto integerLiteral = dyn_cast<IntegerLiteralExpr>(literal)) {
+    return buildBuiltinLiteralArgs(SGF, C, integerLiteral);
+  } else if (auto floatLiteral = dyn_cast<FloatLiteralExpr>(literal)) {
+    return buildBuiltinLiteralArgs(SGF, C, floatLiteral);
+  } else {
+    return buildBuiltinLiteralArgs(
+        SGF, C, dyn_cast<MagicIdentifierLiteralExpr>(literal));
```
Comment (theblixguy) [inline lib/SILGen/SILGenApply.cpp:1918]: I think you should either use `cast` or move it into an if check like above and add a `llvm_unreachable` at the end. This is so we crash here if someone adds a new literal expression but doesn’t update this function.

File: lib/SILGen/SILGenApply.cpp [hunk 2]
```diff
@@ -1786,6 +1786,140 @@ static bool hasUnownedInnerPointerResult(CanSILFunctionType fnType) {
...
+    return buildBuiltinLiteralArgs(SGF, C, booleanLiteral);
+  } else if (auto integerLiteral = dyn_cast<IntegerLiteralExpr>(literal)) {
+    return buildBuiltinLiteralArgs(SGF, C, integerLiteral);
+  } else if (auto floatLiteral = dyn_cast<FloatLiteralExpr>(literal)) {
+    return buildBuiltinLiteralArgs(SGF, C, floatLiteral);
+  } else {
+    return buildBuiltinLiteralArgs(
+        SGF, C, cast<MagicIdentifierLiteralExpr>(literal));
+  }
+  llvm_unreachable("bad literal expr");
```
Comment (theblixguy) [inline lib/SILGen/SILGenApply.cpp:1920]: It’s not needed since the cast above will trigger an assert if one passes a different literal expression.

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test source compatibility

**Comment by Interfere** [comment]
> **Build failed**
> [Swift Test OS X Platform](https://ci.swift.org/job/swift-PR-osx/23886/)
> Git Sha - [4fa17bf](https://github.com/apple/swift/commit/4fa17bf59774ae6543a33b1a29f01320705f4dfe)
Screenshot 2020-10-05 at 12 19 16


> **Build failed**
> [Swift Test Linux Platform](https://ci.swift.org/job/swift-PR-Linux/23895/)
> Git Sha - [4fa17bf](https://github.com/apple/swift/commit/4fa17bf59774ae6543a33b1a29f01320705f4dfe)
Screenshot 2020-10-05 at 12 18 58

@CodaFi Is this a CI related errors, or have I missed something in launching the test fixtures locally? Could you please relaunch? Thanks.

**Comment by xedin** [comment]
That’s a CI related issue. Fixed by updating cmake. Let’s re-run everything.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by Interfere** [comment]
> **Build failed**
> [Swift Test Linux Platform](https://ci.swift.org/job/swift-PR-Linux/23895/)
> Git Sha - [4fa17bf](https://github.com/apple/swift/commit/4fa17bf59774ae6543a33b1a29f01320705f4dfe)
Screenshot 2020-10-05 at 19 34 01

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
I'm just going to land this, we could always revert later if something comes up. Thank you, @Interfere!

**Pull request merged**